WordPress/WordPressStatsWidgets/LockScreenWidgets/LockScreenSiteListProvider.swift

```diff
@@ -0,0 +1,93 @@
+import WidgetKit
+import SwiftUI
+
+struct LockScreenSiteListProvider<T: HomeWidgetData & LockScreenStatsWidgetData>: IntentTimelineProvider {
+
+    let service: StatsWidgetsService
+    let placeholderContent: T
+
+    // refresh interval of the widget, in minutes
+    let refreshInterval = 30
+    // minimum elapsed time, in minutes, before new data are fetched from the backend.
+    let minElapsedTimeToRefresh = 1
+
+    private var defaultSiteID: Int? {
+
+        UserDefaults(suiteName: WPAppGroupName)?.object(forKey: AppConfiguration.Widget.Stats.userDefaultsSiteIdKey) as? Int
+    }
+
+    private let widgetDataLoader = WidgetDataReader<T>()
+
+    func placeholder(in context: Context) -> LockScreenStatsWidgetEntry {
+        LockScreenStatsWidgetEntry.siteSelected(placeholderContent, context)
+    }
+
+    func getSnapshot(for configuration: SelectSiteIntent, in context: Context, completion: @escaping (LockScreenStatsWidgetEntry) -> Void) {
+
+        guard let site = configuration.site,
+              let siteIdentifier = site.identifier,
+              let widgetData = widgetDataLoader.widgetData(for: siteIdentifier) else {
+
+            if let siteID = defaultSiteID, let content = T.read()?[siteID] {
+                completion(.siteSelected(content, context))
+            } else {
+                completion(.siteSelected(placeholderContent, context))
+            }
+            return
+        }
+
+        completion(.siteSelected(widgetData, context))
+    }
+
+    func getTimeline(for configuration: SelectSiteIntent, in context: Context, completion: @escaping (Timeline<LockScreenStatsWidgetEntry>) -> Void) {
```

Comment:
There're many little details in this logic crafted over time. Both logic checks as well as caching functionality. 🤔

Do you see any potential risks to having this (almost) duplicated? If yes, do those risks merit having a separate provider?

Reply:
Yes, it's absolutely a bad smell. Many details are highly duplicated I just copied from another provider. It's a high risk to cause inconsistent behavior and not easy to read because it's mixed with different level logic.

In my opinion, the timeline provider is the component closest to the use case, but some details, such as the cache, are obviously not at the same level. I assume these low-level logic components should rarely change when introducing a new timeline provider for the new widgets.

Extracting them to a standalone data reader could protect them from easy changes and make them testable due to no dependencies on WidgetKit and SwiftUI.